### PR TITLE
EVG-14366 Allow FindPatchesByProject to search by ProjectIdentifiers

### PR DESCRIPTION
**Pull request by bynn**
[EVG-14366](https://jira.mongodb.org/browse/EVG-14366)

### Description 
FindPatchesByProject can now search even if it's given an identifier

### Testing 
unittest

### CODE REVIEW

File: rest/data/patch.go [hunk 1]
```diff
@@ -342,7 +346,7 @@ func (hp *MockPatchConnector) FindPatchesByProject(projectId string, ts time.Tim
 	}
 	for i := len(hp.CachedPatches) - 1; i >= 0; i-- {
 		p := hp.CachedPatches[i]
-		if *p.ProjectId == projectId && !p.CreateTime.After(ts) {
+		if (*p.ProjectId == projectId || *p.ProjectIdentifier == projectId) && !p.CreateTime.After(ts) {
```
Comment (ablack12) [inline rest/data/patch.go:349]: instead can we can just check if *p.ProjectId == id

**Review by ablack12 — CHANGES_REQUESTED** [review]

File: rest/data/patch.go (hunk 1 above)
Comment (bynn) [reply rest/data/patch.go]: done

File: rest/data/patch.go [hunk 2]
```diff
@@ -327,10 +331,20 @@ func (p *DBPatchConnector) GetPatchRawPatches(patchID string) (map[string]string
...
-	CachedPriority   map[string]int64
-	CachedRawPatches map[string]string
+	CachedPatches     []restModel.APIPatch
+	CachedProjectRefs []restModel.APIProjectRef
+	CachedAborted     map[string]string
+	CachedPriority    map[string]int64
+	CachedRawPatches  map[string]string
+}
+
+func (hp *MockPatchConnector) GetIdForProject(identifier string) (string, error) {
```
Comment (ablack12) [inline rest/data/patch.go:341]: I don't think you should make a separate connector function for this. The mock functions are meant to mock the function of the same name for the _real_ connector, and GetIdForProject is part of the FindPatchesByProject connector function, not its own function. I would embed this logic below.

**Review by ablack12 — CHANGES_REQUESTED** [review]
almost there!

File: rest/data/patch.go (hunk 2 above)
Comment (bynn) [reply rest/data/patch.go]: done

File: rest/data/patch.go [hunk 3]
```diff
@@ -39,11 +39,18 @@ type DBPatchConnector struct{}
 // FindPatchesByProject uses the service layer's patches type to query the backing database for
 // the patches.
 func (pc *DBPatchConnector) FindPatchesByProject(projectId string, ts time.Time, limit int) ([]restModel.APIPatch, error) {
-	patches, err := patch.Find(patch.PatchesByProject(projectId, ts, limit))
+	apiPatches := []restModel.APIPatch{}
+	id, err := model.GetIdForProject(projectId)
+	if id == "" {
```
Comment (ablack12) [inline rest/data/patch.go:44]: this will also be true in the error case, so we should check the user first

File: rest/data/patch.go [hunk 4]
```diff
@@ -340,9 +348,18 @@ func (hp *MockPatchConnector) FindPatchesByProject(projectId string, ts time.Tim
 	if limit <= 0 {
 		return patchesToReturn, nil
 	}
+	var id string
+	for _, p := range hp.CachedProjectRefs {
+		if *p.Id == projectId || *p.Identifier == projectId {
+			id = *p.Id
+		}
+	}
+	if id == "" {
```
Comment (ablack12) [inline rest/data/patch.go:357]: should error here

**Review by ablack12 — CHANGES_REQUESTED** [review]

File: rest/data/patch.go (hunk 4 above)
Comment (bynn) [reply rest/data/patch.go:354]: done

File: rest/data/patch.go (hunk 3 above)
Comment (bynn) [reply rest/data/patch.go]: done

**Comment by bynn** [comment]
This ticket changes the behavior of the api because it now returns an error when projects don't exist when it previously just returned 200 with StatusOk

File: rest/data/patch_test.go [hunk 5]
```diff
@@ -144,8 +163,13 @@ func (s *PatchConnectorFetchByProjectSuite) TestFetchTooFew() {
...
 }
 
-func (s *PatchConnectorFetchByProjectSuite) TestFetchNonexistentFail() {
-	patches, err := s.ctx.FindPatchesByProject("zzz", s.time, 1)
+func (s *PatchConnectorFetchByProjectSuite) TestProjectNonexistentFail() {
+	_, err := s.ctx.FindPatchesByProject("zzz", s.time, 1)
+	s.Error(err)
+}
+
+func (s *PatchConnectorFetchByProjectSuite) TestPatchNonexistentFail() {
```
Comment (ablack12) [inline rest/data/patch_test.go:171]: "PatchNonexistentFail" seems wrong; we want it to succeed. Maybe "EmptyPatchesOkay" ?

**Review by ablack12 — APPROVED** [review]
LGTM with nit. (the failing batchtime test is just occasionally flakey so you can ignore it)

**Review by malikchaya2 — APPROVED** [review]

File: rest/data/patch_test.go (hunk 5 above)
Comment (bynn) [reply rest/data/patch_test.go]: done

**Pull request merged**